tensorflow/compiler/tf2tensorrt/convert/op_converter_registry.cc

```diff
@@ -0,0 +1,108 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include "tensorflow/compiler/tf2tensorrt/convert/op_converter_registry.h"
+
+#include <set>
+#include <utility>
+
+#include "tensorflow/core/platform/mutex.h"
+
+#if GOOGLE_CUDA && GOOGLE_TENSORRT
+
+namespace tensorflow {
+namespace tensorrt {
+namespace convert {
+
+// Every op converter function is associated with a Tensorflow operation which
+// it converts. When multiple converters are registered to the same Tensorflow
+// op, the converter with the lowest priority takes precedence.
+using OpConverterRegistration = std::pair<int, OpConverter>;
+struct OpConverterRegistrationComparison {
+  bool operator()(const OpConverterRegistration& lhs,
+                  const OpConverterRegistration& rhs) const {
+    return lhs.first < rhs.first;
+  }
+};
+
+class OpConverterRegistry::Impl {
+ public:
+  ~Impl() = default;
+
+  InitOnStartupMarker Register(const string& name, const int priority,
+                               OpConverter converter) {
+    mutex_lock lock(mu_);
+    auto set = registry_.find(name);
+    if (set == registry_.end()) {
+      std::multiset<OpConverterRegistration, OpConverterRegistrationComparison>
+          new_set;
+      new_set.insert({priority, converter});
+      registry_.insert({name, new_set});
+    } else {
+      set->second.insert({priority, converter});
+    }
+    return {};
+  }
+
+  StatusOr<OpConverter> LookUp(const string& name) {
+    mutex_lock lock(mu_);
+    auto found = registry_.find(name);
+    if (found != registry_.end()) {
+      auto item = *found->second.begin();
+      return item.second;
+    }
+    return errors::NotFound("No converter for op ", name);
+  }
+
+  void Clear(const std::string& name) {
+    mutex_lock lock(mu_);
+    auto itr = registry_.find(name);
+    if (itr == registry_.end()) {
+      return;
+    }
+    registry_.erase(itr);
+  }
+
+ private:
+  mutable mutex mu_;
+  mutable std::unordered_map<
+      std::string,
+      std::multiset<OpConverterRegistration, OpConverterRegistrationComparison>>
+      registry_ TF_GUARDED_BY(mu_);
```

Comment:
I think we should first consider the following to see if we want to simplify this data structure:
(1) Do we want to record multiple converters for an OP?
(2) Do we want to record multiple converters with the same priority for an OP?

We should never record the converters that we don't actually use.

For (2), I tend to believe the answer is NO.
For (1), it depends on whether we want to do this:
      if the conversion with a higher priority converter fails, call try the converters with lower priority.
I tend to believe doing the above is NOT necessary. If a user want to provide a converter but only handle some cases, they should explicitly call the "default converter" inside their converter.

If we agree on the above, then we only need an unordered_map from "string" to "(priority, OpConverter)". The registration routine should issue a warning message when it sees a converter with higher priority overwrites another converter with lower priority.

Reply:
So to summarize, are you saying:

- Change `mutliset` to `set`
- Change sort so higher "priority number" takes precedence rather than lower, which is how it is now
- Issue a warning when an override occurs?

Reply:
If we will only keep one converter with the highest priority we have seen for each op, we only need to maintain a map from op-name to (priority, converter). Not need to use a set to host (priority, converter). Right?

Reply:
Okay, will do

Reply:
fixed